### PR TITLE
New endpoints

### PR DESCRIPTION
**Pull request by sasurobert**
started implementation for new endpoints and activation flag

### CODE REVIEW

**Review by matei-p — DISMISSED** [review]

File: arwen/cryptoapi/cryptoei.go [hunk 1]
```diff
@@ -340,6 +425,32 @@ func v1_4_ripemd160(context unsafe.Pointer, dataOffset int32, length int32, resu
 	return 0
 }
 
+//export v1_4_managedRipemd160
+func v1_4_managedRipemd160(context unsafe.Pointer, inputHandle int32, outputHandle int32) int32 {
+	runtime := arwen.GetRuntimeContext(context)
+	managedType := arwen.GetManagedTypesContext(context)
+	crypto := arwen.GetCryptoContext(context)
+	metering := arwen.GetMeteringContext(context)
```
Comment (andreibancioiu) [inline arwen/cryptoapi/cryptoei.go:433]: Sometimes `StartGasTracing` is used, while other times `UseGasAndAddTracedGas` is used. Is this all right?

File: arwen/elrondapi/managedei.go [hunk 2]
```diff
@@ -926,3 +957,102 @@ func v1_4_managedTransferValueExecute(
...
+//export v1_4_managedBufferToHex
+func v1_4_managedBufferToHex(context unsafe.Pointer, sourceHandle int32, destHandle int32) {
+	runtime := arwen.GetRuntimeContext(context)
+	metering := arwen.GetMeteringContext(context)
+	managedType := arwen.GetManagedTypesContext(context)
+
+	gasToUse := metering.GasSchedule().ManagedBufferAPICost.MBufferSetBytes
+	metering.UseGasAndAddTracedGas(managedBufferToHexName, gasToUse)
+
+	mBuff, err := managedType.GetBytes(sourceHandle)
```
Comment (andreibancioiu) [inline arwen/elrondapi/managedei.go:1050]: In this case, shouldn't we perhaps use a little bit of gas with respect to the size of the buffer, as well?

File: arwen/cryptoapi/cryptoei.go [hunk 3]
```diff
@@ -980,12 +1458,54 @@ func v1_4_unmarshalCompressedEC(
 		return int32(len(data))
 	}
 
+	return commonUnmarshalCompressedEC(context, xResultHandle, yResultHandle, ecHandle, data)
```
Comment (andreibancioiu) [inline arwen/cryptoapi/cryptoei.go:1461]: Can you confirm this change is backwards compatible? (i.e., consensus will work on this kind of errors, before the new features get activated).

**Review by andreibancioiu — APPROVED** [review]

File: arwen/cryptoapi/cryptoei.go [hunk 4]
```diff
@@ -383,6 +494,49 @@ func v1_4_verifyBLS(
...
+	crypto := arwen.GetCryptoContext(context)
+	metering := arwen.GetMeteringContext(context)
+	managedType := arwen.GetManagedTypesContext(context)
+	metering.StartGasTracing(verifyBLSName)
+
+	gasToUse := metering.GasSchedule().CryptoAPICost.VerifyBLS
+	metering.UseAndTraceGas(gasToUse)
+
+	keyBytes, err := managedType.GetBytes(keyHandle)
+	if arwen.WithFault(err, context, runtime.ManagedBufferAPIErrorShouldFailExecution()) {
```
Comment (camilbancioiu) [inline arwen/cryptoapi/cryptoei.go:514]: `WithFaultIfFailAlwaysActive()`? Not sure right now, might be ok as is.

File: arwen/cryptoapi/cryptoei.go [hunk 5]
```diff
@@ -540,6 +791,35 @@ func v1_4_encodeSecp256k1DerSignature(
...
+) int32 {
+	runtime := arwen.GetRuntimeContext(context)
+	crypto := arwen.GetCryptoContext(context)
+	metering := arwen.GetMeteringContext(context)
+	managedType := arwen.GetManagedTypesContext(context)
+
+	gasToUse := metering.GasSchedule().CryptoAPICost.EncodeDERSig
+	metering.UseGasAndAddTracedGas(encodeSecp256k1DerSignatureName, gasToUse)
+
+	r, err := managedType.GetBytes(rHandle)
```
Comment (camilbancioiu) [inline arwen/cryptoapi/cryptoei.go:807]: Consume gas.

File: arwen/cryptoapi/cryptoei.go [hunk 6]
```diff
@@ -540,6 +791,35 @@ func v1_4_encodeSecp256k1DerSignature(
...
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	s, err := managedType.GetBytes(sHandle)
```
Comment (camilbancioiu) [inline arwen/cryptoapi/cryptoei.go:812]: Consume gas.

File: arwen/cryptoapi/cryptoei.go [hunk 7]
```diff
@@ -713,6 +993,50 @@ func v1_4_scalarBaseMultEC(
...
+	managedType := arwen.GetManagedTypesContext(context)
+	metering.StartGasTracing(scalarBaseMultECName)
+
+	curveMultiplier := managedType.GetScalarMult100xCurveGasCostMultiplier(ecHandle)
+	if curveMultiplier < 0 {
+		_ = arwen.WithFault(arwen.ErrNoEllipticCurveUnderThisHandle, context, runtime.CryptoAPIErrorShouldFailExecution())
+		return 1
+	}
+
+	data, err := managedType.GetBytes(dataHandle)
```
Comment (camilbancioiu) [inline arwen/cryptoapi/cryptoei.go:1018]: Consume gas for load.

File: arwen/cryptoapi/cryptoei.go [hunk 8]
```diff
@@ -808,94 +1182,156 @@ func v1_4_marshalEC(
...
+	context unsafe.Pointer,
+	xPairHandle int32,
+	yPairHandle int32,
+	ecHandle int32,
+	resultHandle int32,
+) int32 {
+	runtime := arwen.GetRuntimeContext(context)
+	result, err := commonMarshalEC(context, xPairHandle, yPairHandle, ecHandle)
+	if err != nil {
+		_ = arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution())
```
Comment (camilbancioiu) [inline arwen/cryptoapi/cryptoei.go:1209]: `WithFaultIfFailIsAlwaysActive()`?

File: arwen/cryptoapi/cryptoei.go (hunk 1 above)
Comment (sasurobert) [reply arwen/cryptoapi/cryptoei.go:433]: yes. both can be used

File: arwen/cryptoapi/cryptoei.go (hunk 3 above)
Comment (sasurobert) [reply arwen/cryptoapi/cryptoei.go:1461]: yes. it is.

File: arwen/elrondapi/managedei.go (hunk 2 above)
Comment (sasurobert) [reply arwen/elrondapi/managedei.go:1050]: it is fine - as gas was taken out once when the sourceHandle was created.

File: arwen/cryptoapi/cryptoei.go (hunk 4 above)
Comment (sasurobert) [reply arwen/cryptoapi/cryptoei.go:514]: everyone is returning true. all errors return true.

File: arwen/cryptoapi/cryptoei.go (hunk 5 above)
Comment (sasurobert) [reply arwen/cryptoapi/cryptoei.go:807]: no need. it uses enough.

File: arwen/cryptoapi/cryptoei.go (hunk 6 above)
Comment (sasurobert) [reply arwen/cryptoapi/cryptoei.go:812]: no need. it uses enough.

File: arwen/cryptoapi/cryptoei.go (hunk 7 above)
Comment (sasurobert) [reply arwen/cryptoapi/cryptoei.go:1018]: it is used in 1024

File: arwen/cryptoapi/cryptoei.go [hunk 9]
```diff
@@ -771,8 +1095,58 @@ func v1_4_scalarMultEC(
...
+	managedType := arwen.GetManagedTypesContext(context)
+	metering.StartGasTracing(scalarMultECName)
+
+	curveMultiplier := managedType.GetScalarMult100xCurveGasCostMultiplier(ecHandle)
+	if curveMultiplier < 0 {
+		_ = arwen.WithFault(arwen.ErrNoEllipticCurveUnderThisHandle, context, runtime.CryptoAPIErrorShouldFailExecution())
+		return 1
+	}
+
+	data, err := managedType.GetBytes(dataHandle)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/cryptoapi/cryptoei.go:1122]: it is used on the next lines

File: arwen/cryptoapi/cryptoei.go (hunk 8 above)
Comment (sasurobert) [reply arwen/cryptoapi/cryptoei.go:1209]: done

File: arwen/cryptoapi/cryptoei.go [hunk 10]
```diff
@@ -1093,6 +1647,39 @@ func v1_4_createEC(context unsafe.Pointer, dataOffset int32, dataLength int32) i
...
+//export v1_4_managedCreateEC
+func v1_4_managedCreateEC(context unsafe.Pointer, dataHandle int32) int32 {
+	managedType := arwen.GetManagedTypesContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().CryptoAPICost.EllipticCurveNew
+	metering.UseGasAndAddTracedGas(createECName, gasToUse)
+
+	data, err := managedType.GetBytes(dataHandle)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/cryptoapi/cryptoei.go:1659]: it is done on the next lines

**Review by andreibancioiu — APPROVED** [review]

**Pull request merged**